FirebaseAppDistribution/Sources/FIRAppDistribution.m

```diff
@@ -116,181 +123,160 @@ + (instancetype)appDistribution {
 }
 
-- (void)cleanupUIWindow {
-  if (self.window) {
-    self.window.hidden = YES;
-    self.window = nil;
-  }
+- (NSError *)NSErrorForErrorCodeAndMessage:(FIRAppDistributionError)errorCode
+                                   message:(NSString *)message {
+  NSDictionary *userInfo = @{FIRAppDistributionErrorDetailsKey : message};
+  return [NSError errorWithDomain:FIRAppDistributionErrorDomain code:errorCode userInfo:userInfo];
 }
 
-- (void)handleReleasesAPIResponseWithData:data
-                               completion:(FIRAppDistributionUpdateCheckCompletion)completion {
-  NSError *error = nil;
-  NSDictionary *object = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
-
-  NSArray *releaseList = [object objectForKey:@"releases"];
-  for (NSDictionary *releaseDict in releaseList) {
-    if (![[releaseDict objectForKey:@"latest"] boolValue]) continue;
-
-    NSString *codeHash = [releaseDict objectForKey:@"codeHash"];
-    FIRAppDistributionMachO *machO =
-        [[FIRAppDistributionMachO alloc] initWithPath:[[NSBundle mainBundle] executablePath]];
-
-    if (![codeHash isEqualToString:machO.codeHash]) {
-      FIRAppDistributionRelease *release =
-          [[FIRAppDistributionRelease alloc] initWithDictionary:releaseDict];
-      dispatch_async(dispatch_get_main_queue(), ^{
-        completion(release, nil);
-      });
+- (NSError *_Nullable)mapFetchReleasesError:(NSError *)error {
+  if ([error domain] == kFIRFADApiErrorDomain) {
+    FIRFADErrorLog(@"Failed to retrieve releases: %ld", (long)[error code]);
+    switch ([error code]) {
+      case FIRFADApiErrorTimeout:
+        return [self NSErrorForErrorCodeAndMessage:FIRAppDistributionErrorNetworkFailure
+                                           message:@"Failed to fetch releases due to timeout."];
+      case FIRFADApiErrorUnauthenticated:
+      case FIRFADApiErrorUnauthorized:
+      case FIRFADApiTokenGenerationFailure:
+      case FIRFADApiInstallationIdentifierError:
+      case FIRFADApiErrorNotFound:
+        return [self NSErrorForErrorCodeAndMessage:FIRAppDistributionErrorAuthenticationFailure
+                                           message:@"Could not authenticate tester"];
+      default:
+        return [self NSErrorForErrorCodeAndMessage:FIRAppDistributionErrorUnknown
+                                           message:@"Failed to fetch releases for unknown reason."];
     }
   }
+
+  FIRFADErrorLog(@"Failed to retrieve releases with unexpected domain %@: %ld", [error domain],
+                 (long)[error code]);
+  return [self NSErrorForErrorCodeAndMessage:FIRAppDistributionErrorUnknown
+                                     message:@"Failed to fetch releases for unknown reason."];
 }
+
+- (void)fetchNewLatestRelease:(FIRAppDistributionUpdateCheckCompletion)completion {
+  [FIRFADApiService
+      fetchReleasesWithCompletion:^(NSArray *_Nullable releases, NSError *_Nullable error) {
```

Comment:
nit: using static methods and singletons usually limits code testability. Injecting an instance of `FIRFADApiService` usually provides much more room for testing and mocking without touching Objective-C runtime as well as improves the code readability.

Reply:
@maksymmalyhin can you expand on that? I have not had issues mocking static methods using OCMock and tend to air on the side of `if it's not holding state, make it a stateless static class`. Certainly a singleton pattern accomplishes the same goal, but I would love to dig in more to the testing impediments you've run into as well as your feelings around consistency and style for my own iOS learning.

Reply:
There are different points of view on it. Some of them are pretty radical like go/srl-dev/why-fakes#no-ocmock which basically means "never use OCMock". From the other side, from my personal experience, opting out of mocking completely slows down development pretty significantly when fakes used by a single test have to be implemented from scratch. But mocking class methods proved to be worse because it affects entire runtime and not limited to your particular test scope which may lead to unexpected interference with other tests or other unexpected issues.

Couple more things to consider for this particular case:
1. `FIRFADApiService` actually access and modifies different states but it does it via e.g. `NSURLSession` and `FIRInstallations` singletons
2. Using class methods makes the dependency on  `FIRFADApiService` more implicit. One has to look through the implementation in order to spot it
3. If the dependency injected, switching between mock and fake is much easier and would require only test code modifications

I think @samedson may provide a bit more details on the topic since Crashlytics team consciously opted out of using `OCMock`  in the product tests.

Reply:
I spoke with @samedson about this more, and I think I agree with both of you. I think given that this will be a bigger refactor, do you mind if we take this on as a subsequent change @maksymmalyhin?

Reply:
sure, not an issue from my end

Reply:
Sweet! Thank you!